integration-tests/smoke: add batching test

makramkd opened the pull request:
Ticket: https://smartcontract-it.atlassian.net/browse/CCIP-4166

Context for CCIP onchain reviewers: We add the Multicall3 contract in tests/helpers so that we can use it to trigger multiple CCIP messages in a single tx in the test. Repo here: https://github.com/mds1/multicall

### Requires
N/A

### Supports
N/A

makramkd committed:
integration-tests/smoke: add batching test

makramkd commented on deployment/ccip/deploy.go:
```diff
@@ -233,7 +237,6 @@ func DeployPrerequisiteContracts(e deployment.Environment, ab deployment.Address
 			return err
 		}
 		lggr.Infow("deployed linkToken", "addr", linkToken.Address)
-		linkTokenContract = linkToken.Contract
```
Linter said this was unused

makramkd commented on deployment/ccip/deploy.go:
```diff
@@ -256,10 +259,28 @@ func DeployPrerequisiteContracts(e deployment.Environment, ab deployment.Address
 			return err
 		}
 		e.Logger.Infow("deployed router", "addr", routerContract.Address)
-		r = routerContract.Contract
```
Unused

makramkd committed:
Merge branch 'develop' into mk/p1/CCIP-4166

makramkd committed:
fix solhint, forge fmt, golangci-lint

makramkd committed:
Merge branch 'develop' into mk/p1/CCIP-4166

makramkd committed:
add batching test

makramkd commented on deployment/ccip/deploy_home_chain.go:
```diff
@@ -260,27 +260,6 @@ func DeployHomeChain(
 
-// getNodeOperatorIDMap returns a map of node operator names to their IDs
-// If maxNops is greater than the number of node operators, it will return all node operators
-func getNodeOperatorIDMap(capReg *capabilities_registry.CapabilitiesRegistry, maxNops uint32) (map[string]uint32, error) {
```
Unused

makramkd committed:
add single source batching test case

makramkd commented on contracts/src/v0.8/ccip/test/helpers/Multicall3.sol:
```diff
@@ -0,0 +1,245 @@
+// SPDX-License-Identifier: MIT
+// cribbed from https://github.com/mds1/multicall/commit/ebd8b64457454fc10037b3a3ea858f9c08dad4d3
+// changed solidity version to 0.8.24 from 0.8.12
+// ran forge fmt
+// solhint-disable
```
Required to fix Solidity/Forge CI

makramkd committed:
rm unused func

AnieeG commented on deployment/ccip/deploy.go:
```diff
@@ -262,10 +265,28 @@ func DeployPrerequisiteContracts(e deployment.Environment, ab deployment.Address
 	} else {
 		e.Logger.Infow("router already deployed", "addr", chainState.Router.Address)
 	}
+	if mc3 == nil {
```
Do we need to deploy mc3 always? Or should it have some conditional input?

makramkd replied:
What's the best way to condition it?

mateusz-sekara replied:
We use `DeployCCIPContractConfig` to tell the deployer whether to deploy or not USDC-related contracts.
https://github.com/smartcontractkit/chainlink/blob/43f0ff1334b2e0c12b5073125dc524874839b70f/deployment/ccip/deploy.go#L339

Not sure if it fits your case. Another way would be to deploy multicall explicitly on demand (using a separate function)

```go
ccdeploy.DeployMultiCall(chainId)
```

AnieeG commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,299 @@
+				nil,
+				ccipocr3.NewSeqNumRange(startSeqNum[sourceChain1], endSeqNum[sourceChain1]),
+			)
+			require.NoErrorf(t, err, "failed to confirm commit from chain %d", sourceChain1)
```
please note that a require failure here won't fail the actual test immediately. The test will be reported in failed status but after all go routine is completed.
If you want to fail the test immediately for any failure in individual routines please consider using errgroup.

makramkd replied:
Are you sure? `require` calls `t.FailNow` which should shut down the whole test

jmank88 replied:
The fatal variants of the `t.Testing` methods are not supposed to be called from any goroutine other than the main one the test began on because the mechanism relies on `runtime.Goexit`:
>  FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test. Calling FailNow does not stop those other goroutines.

https://pkg.go.dev/testing#T.FailNow

makramkd replied:
Noice, alright will fix this then.

makramkd replied:
Fixed

AnieeG commented on deployment/ccip/test_helpers.go:
```diff
@@ -268,6 +274,28 @@ func CCIPSendRequest(
+// CCIPSendCalldata packs the calldata for the Router's ccipSend method.
+// This is expected to be used in Multicall scenarios (i.e multiple ccipSend calls
+// in a single transaction).
+// Multicalls can't be made with fee token 0x0 due to the Multicall itself not
```
Are you not sending the value as fee here in `genMessages` instead of using feetoken - https://github.com/smartcontractkit/chainlink/pull/15292/files#diff-a0967558956b7b4a235492bf27d9cb41b7b7a2e0d05b9cab4f9e2871b0893942R294 ?

makramkd replied:
Oops yeah this is a stale comment, good catch

makramkd committed:
update comment

makramkd committed:
Merge branch 'develop' into mk/p1/CCIP-4166

makramkd committed:
don't call require in goroutines

mateusz-sekara commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,295 @@
+		)
+
+		wg.Add(1)
+		go func(sourceChainSelector uint64) {
```
That could be extracted to a separate function and then reused for every chain

```go
func sendMessageAsync(sourceChain, wg) {
                         defer wg.Done()
			err := sendMessages(
				ctx,
				t,
				e.Env.Chains[sourceChainSelector],
				e.Env.Chains[sourceChainSelector].DeployerKey,
				state.Chains[sourceChainSelector].OnRamp,
				state.Chains[sourceChainSelector].Router,
				state.Chains[sourceChainSelector].Multicall3,
				destChain,
				numMessages,
				common.LeftPadBytes(state.Chains[destChain].Receiver.Address().Bytes(), 32),
			)
			// you can just log the error and fail?
                        // or use the chan for sending errs
}


for soureChain := range []uint64{sourceChain1, sourceChain2} {
  wg.add(1)
  go sendMessageAsync(sourceChain, errs, wg)
}
wg.wait()

makramkd replied:
Fixed

mateusz-sekara commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,295 @@
+		errs = nil
+
+		wg.Add(1)
+		go func() {
```
Similarly to the previous one, this involves a lot of duplication. I guess it might be very useful for other tests, so might be useful to extract that to some global function that accepts multiple sources and waits for them in async manner

0xnogo commented on deployment/ccip/deploy.go:
```diff
@@ -260,10 +263,28 @@ func DeployPrerequisiteContracts(e deployment.Environment, ab deployment.Address
+	if mc3 == nil {
+		multicall3Contract, err := deployment.DeployContract(e.Logger, chain, ab,
+			func(chain deployment.Chain) deployment.ContractDeploy[*multicall3.Multicall3] {
+				multicall3Addr, tx2, multicall3Wrapper, err2 := multicall3.DeployMulticall3(
```
seems like this error in not checked

makramkd replied:
I believe its checked downstream (within `DeployContract`)

mateusz-sekara commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,295 @@
+	require.NoError(t, ccdeploy.AddLaneWithDefaultPrices(e.Env, state, sourceChain2, destChain))
+
+	const (
+		numMessages = 5
```
5 is very little, shouldn't we use some higher value? 20 or 30?

mateusz-sekara replied:
If the delay caused by the sending ccip messages leads to fragmentation (creating multiple batches), you can try sending messages **before** proposing job specs. That way you "delay" the off-chain processing and gives chain more time to catch up

makramkd replied:
It shouldn't lead to fragmentation for commit because these messages are created in the same tx

makramkd replied:
I can bump this to 30 and see what happens.

makramkd replied:
Commit works fine with 30 but exec doesn't, there's a bug there due to that. We have two options here: either drop this to work with exec, or keep it at 30 and only assert commit, while we try to figure out how to properly address this in exec

mateusz-sekara commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,295 @@
+			nil,
+			ccipocr3.NewSeqNumRange(startSeqNum[sourceChain1], endSeqNum[sourceChain1]),
+		)
+		require.NoErrorf(t, err, "failed to confirm commit from chain %d", sourceChain1)
```
Even though the test is oriented around commit batching, I think it would be worth waiting for messages execution. We don't care about batch sizes for exec, but we want to make sure everything is successfully executed (e.g. this large commit batch is properly processed by the exec)

makramkd replied:
Yeah, I thought about that as well. Exec doesn't have to batch like commit due to gas limits though

mateusz-sekara commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,295 @@
+		// that one report.
+		require.Equal(t, sourceChain1Report, sourceChain2Report, "commit reports should be the same")
+
+		startSeqNum[sourceChain1] = endSeqNum[sourceChain1] + 1
```
It's not used anywhere, is it?

makramkd replied:
This isn't used anywhere but if we add test cases further down (which was the plan, wanted to add something where we manual exec a message that is part of a root with multiple msgs)

asoliman92 replied:
If it's not gonna be used in this PR then better remove them to make the next PR easier to understand when seeing it added.

mateusz-sekara commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,295 @@
+		go func() {
+			defer wg.Done()
+			var err error
+			sourceChain2Report, err = ccdeploy.ConfirmCommitWithExpectedSeqNumRange(t,
```
Does this check enforce all seqNrs to be in the same root?

makramkd replied:
Yeah, believe so, it checks that the root min seq nr >= the given min seq nr and max seq nr <= the given max seq nr

asoliman92 replied:
I double checked and it does so 👍

0xnogo commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,295 @@
+		wg.Wait()
+
+		for _, err := range errs {
+			require.NoError(t, err)
```
I think this will not stop the execution of the test as errors from goroutines are only checked after all goroutines finish. Should we do something like:

```
for err := range errsChan {
    if err != nil {
        t.Fatalf("Error occurred: %v", err)
    }
}
```

makramkd replied:
Not sure what you mean, `require.NoError` will fail the whole test if the err is not nil

0xnogo commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,295 @@
+	}
+
+	// check that the message was emitted
+	iter, err := sourceOnRamp.FilterCCIPMessageSent(
```
nit Should we close the iterator?

0xnogo commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,295 @@
+
+		// the reports should be the same for both, since both roots should be batched within
+		// that one report.
+		require.Equal(t, sourceChain1Report, sourceChain2Report, "commit reports should be the same")
```
Can the report be both nil?

We could do this before this assertion to avoid this scenario:
```
require.NotNil(t, sourceChain1Report)
require.NotNil(t, sourceChain2Report)
```

asoliman92 replied:
I don't think it should happen because it will be an error then, but extra safety is good as well.

makramkd replied:
They can't be nil, the `ConfirmCommitWithExpectedSeqNumRange` assertion only returns a nil report if there's an error

asoliman92 commented on contracts/src/v0.8/ccip/test/helpers/Multicall3.sol:
```diff
@@ -0,0 +1,245 @@
+// ran forge fmt
+// solhint-disable
+pragma solidity 0.8.24;
+
```
Maybe Rename it to `MulticallAggregator`? What's the 3 for exactly?

makramkd replied:
See PR description and comment above these lines, this is a 3rd party contract

asoliman92 commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,295 @@
+		}
+		endSeqNum map[uint64]ccipocr3.SeqNum = map[uint64]ccipocr3.SeqNum{
+			sourceChain1: ccipocr3.SeqNum(numMessages),
+			sourceChain2: ccipocr3.SeqNum(numMessages),
```
I think it's worth having the following for test completeness:
1. A third chain with no messages
2.  Different numMessages per chain

makramkd replied:
Good ideas - will add these in follow ups once exec is more bulletproof. For (1) it seems like a good followup

asoliman92 commented on deployment/ccip/test_assertions.go:
```diff
@@ -268,17 +269,17 @@ func ConfirmCommitWithExpectedSeqNumRange(
+							t.Logf("Received commit report for [%d, %d] on selector %d from source selector %d expected seq nr range %s, token prices: %v, tx hash: %s",
+								mr.MinSeqNr, mr.MaxSeqNr, dest.Selector, src.Selector, expectedSeqNumRange.String(), event.PriceUpdates.TokenPriceUpdates, event.Raw.TxHash.String())
+							return event, nil
 						}
```
should we return nil, err if no events were found here?

makramkd replied:
No, test should fail

asoliman92 commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,295 @@
+			Data:         []byte(fmt.Sprintf("hello world %d", i)),
+			TokenAmounts: nil,
+			FeeToken:     common.HexToAddress("0x0"),
+			ExtraArgs:    nil,
```
Can we add some random extraArgs here? Will be good for any regression as well.

asoliman92 replied:
I think we should add tokens to the test. It might not add a lot, but it would make the test more comprehensive and ensure a full message with all components works as expected. I noticed @0xsuryansh  used some helper functions for token sending in his PR that we could probably reuse here. We don’t need to verify the token transfers since Suryansh’s tests already cover that, but including tokens would still make the test more complete.

makramkd replied:
Yeah, waiting for the tokens test to merge so I can use the helpers from there

makramkd committed:
Merge branch 'develop' into mk/p1/CCIP-4166

makramkd committed:
fix

makramkd committed:
some fixes

makramkd committed:
Merge branch 'develop' into mk/p1/CCIP-4166

makramkd committed:
rm from mq

makramkd committed:
fix ConfirmExecWithSeqNrs

makramkd committed:
bump msgs to 30, fix multisource concurrency

makramkd committed:
close iter

makramkd committed:
add exec check

bring down the numMessages to 5 again because exec is exceeding the max
observation size

asoliman92 commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -177,6 +191,49 @@ func Test_CCIPBatching(t *testing.T) {
+
+		i = 0
+		var execStates []map[uint64]int
+	outer3:
```
```
for i < len(sourceChains) {
	select {
	case outputErr := <-execErrs:
		require.NoError(t, outputErr.err)
		execStates = append(execStates, outputErr.output)
		i++
	case <-ctx.Done():
		require.FailNow(t, "didn't get all exec reports before test context was done")
	}
}
```

makramkd replied:
That doesn't work

makramkd replied:
Actually that does work, just weirder :D

asoliman92 replied:
I tested it before writing it :D. I know it works but why is it weirder? I think it's much clearer that way.

asoliman92 commented on integration-tests/smoke/ccip_batching_test.go:
```diff
@@ -0,0 +1,428 @@
+		wg.Wait()
+
+		var i int
+	outer:
```
This is duplicated 3 times, please add a function that takes the channel as an input

makramkd replied:
I think having the function here is not as good, there's a ton of vars to take in. Will have to think about how to refactor this to be prettier later

asoliman92 replied:
It's 3 variables
* testing.T
* errs channel
* sourceChains length

Much better IMO than current.

makramkd replied:
The chan types are all different unfortunately, there is `chan error`, `chan outputErr[CommitReport]` and `chan outputErr[map[...]]`

AnieeG commented on deployment/ccip/changeset/deploy.go:
```diff
@@ -261,10 +264,28 @@ func DeployPrerequisiteContracts(e deployment.Environment, ab deployment.Address
 	} else {
 		e.Logger.Infow("router already deployed", "addr", chainState.Router.Address)
 	}
+	if mc3 == nil {
```
TODO need to update this based on deployOpts here https://github.com/smartcontractkit/chainlink/pull/15349/files

AnieeG reviewed (commented):
LGTM. do you want to wait for the deployOpts ?

makramkd committed:
Merge branch 'develop' into mk/p1/CCIP-4166

makramkd committed:
simplify loops

makramkd committed:
Merge branch 'develop' into mk/p1/CCIP-4166

makramkd committed:
fixes

makramkd committed:
move multicall3 to vendor, shared

RensR reviewed (dismissed):
Approving only the Solidity / gethwrapper part of this PR

makramkd committed:
some more fixes
